File: packages/codex-session-relay/src/codex_session_relay/hostloss.py
"""Noticing that the host lost a turn it accepted, and recovering the obligation once (CRW-224).

A dispatch is a transport fact: turn/start returned a turn id. It is not proof that the host still
has that turn. CRW-124's H0 rehearsal killed the App Server right after it accepted turn/start for a
completion to the parent; the rollout kept only task_started, the restarted host listed no such
turn, and the relay held the delivery as dispatched_awaiting_ack for ever. Nothing read the
parent's turns, so the child's report was lost silently and looked exactly like a lost ACK.

The reading here asks the recipient's own turn list for the dispatched turn among the turns begun
since the send (hostadapter.find_in_listing). A loss is concluded only from the host's answer, only
once the send is older than the allowance a whole-second start time and a clock skew need, and only
when a scan of every item of the turns begun since the send shows this attempt's token is not
there (hostadapter.find_token_in): a found token means the message reached the recipient whatever
happened to the turn row, and a scan that stopped at its bound has shown nothing (I-42). A lost
ACK keeps its turn in the list, interrupted, with the message in it, so it reads present and its
path does not change.

A reading that cannot reach an answer by waiting - the listing never reached the send, the listing
is empty once the send is past the allowance, the token scan could not cover the turns since it,
the attempt has no send time - is recorded on the attempt as turn_check_undecided:<reason>, which
status names, rather than left to look like an ordinary wait. It is cleared when a later reading
decides.

Recovery is the ordinary claim path, once. The delivery goes back to queued and the next attempt of
the same event is sent under a new request id; if the host loses that turn too, the delivery is
held under host_lost_turn and nothing sends it again. Only a completion is recovered: it is the one
kind waiting for an acknowledgement, and a revision's dispatch turn is its generation's anchor,
which is never rebound. Other kinds are read and reported, and changed in nothing.
"""

import json
from datetime import datetime

from .delivery import COMPLETION
from .hostadapter import DISPATCH_TURN_SKEW_SECONDS, TURN_PRESENT, TURN_START_PRECISION_SECONDS
from .hostadapter import ListingBounded, ListingEmpty
from .policy import HOST_LOST_TURN, TURN_CHECK_UNDECIDED
from .transport import DISPATCHED, QUEUED, assert_attempt_invariants

PRESENT = "present"
UNKNOWN = "unknown"
# A turn listed in one of these is persisted: a later restart reads it back, interrupted at worst,
# so it cannot be lost any more and need not be read again.
TERMINAL = ("completed", "interrupted", "failed")
TOKEN_SCAN_LIMIT = 200
# The redelivery a loss leads to, or why none followed.
REQUEUED = "queued"
HELD = "held"
NOT_MOVED = "not_moved"
REPORT_ONLY = "report_only"
# Why an unknown reading will not resolve by waiting. These are recorded by name; every other
# unknown - an unreadable host, a send too recent - is simply read again later.
LISTING_BOUNDED = "listing_bounded"
LISTING_EMPTY = "listing_empty"
TOKEN_SCAN_BOUNDED = "token_scan_bounded"
NO_SEND_TIME = "no_send_time"
NO_TURN = "no_turn"
UNDECIDED_MARK = TURN_CHECK_UNDECIDED + ":"


class _Raced(Exception):
    """The attempt changed between the delivery's guarded update and its own."""


def _epoch(stamp):
    try:
        return datetime.fromisoformat(stamp).timestamp()
    except (TypeError, ValueError):
        return None


def read_recipient_turn(adapter, clock, attempt, delivery, turn_id) -> dict:
    """What the recipient's own turns say about this attempt's turn. Reads only; writes nothing.

    finding is present, host_lost_turn or unknown. unknown is not an answer, and detail says why
    no answer was reached: an adapter that cannot list turns, an attempt without a send time, an
    unreadable host, or a send too recent to call its turn lost.
    undecided names the reasons waiting will not fix (listing_bounded, listing_empty,
    token_scan_bounded, no_send_time, no_turn), and is None otherwise.
    """
    reading = {"turnId": turn_id, "finding": UNKNOWN, "status": None, "detail": None,
               "undecided": None}
    thread = delivery["recipient_thread_id"]
    lookup = getattr(adapter, "find_dispatched_turn", None)
    token_since = getattr(adapter, "find_token_since", None)
    if lookup is None or token_since is None:
        reading["detail"] = "this host adapter cannot list the recipient's turns"
        return reading
    if not turn_id:
        reading.update(detail="the attempt names no turn", undecided=NO_TURN)
        return reading
    # When the send STARTED. observed_at is its settlement, later than the send, and would move
    # the cutoff the wrong way; an attempt without the stamp is left unjudged instead.
    sent_at = _epoch(attempt["sent_at"])
    if sent_at is None:
        reading.update(detail="the attempt has no send time, so absence cannot be bounded",
                       undecided=NO_SEND_TIME)
        return reading
    allowance = TURN_START_PRECISION_SECONDS + DISPATCH_TURN_SKEW_SECONDS
    try:
        presence = lookup(thread, turn_id, sent_at=sent_at)
    except ListingBounded as error:
        reading.update(detail=f"undecided: {error}", undecided=LISTING_BOUNDED)
        return reading
    except ListingEmpty as error:
        # A parent whose only turn was the lost delivery lists nothing, and will go on listing
        # nothing: name it rather than read it as a transient failure for ever (review 4). Not
        # before the allowance, when an empty list may still be the host catching up.
        if clock.now() < sent_at + allowance:
            reading["detail"] = (
                f"the recipient lists no turns yet, and the send is less than {allowance:.0f} s "
                f"old: {error}"
            )
        else:
            reading.update(detail=f"undecided: {error}", undecided=LISTING_EMPTY)
        return reading
    except Exception as error:  # noqa: BLE001 - an unreadable host is its own answer
        reading["detail"] = f"unreadable: {type(error).__name__}: {error}"
        return reading
    if presence.finding == TURN_PRESENT:
        reading.update(finding=PRESENT, status=presence.turn.status,
                       detail=f"the recipient lists this turn ({presence.turn.status})")
        return reading
    if clock.now() < sent_at + allowance:
        reading["detail"] = (
            f"the recipient does not list this turn yet ({presence.stop}), but the send is less "
            f"than {allowance:.0f} s old, too recent to call the turn lost"
        )
        return reading
    try:
        scan = token_since(thread, attempt["request_id"], turns=presence.seen + (turn_id,),
                           limit=TOKEN_SCAN_LIMIT)
    except Exception as error:  # noqa: BLE001
        reading["detail"] = (
            f"the recipient does not list this turn, and its items could not be read for this "
            f"attempt's token: {type(error).__name__}: {error}"
        )
        return reading
    if scan.found:
        reading.update(
            finding=PRESENT,
            detail=f"the recipient does not list this turn, but this attempt's token is in its "
                   f"items (turn {scan.turn_id})",
        )
        return reading
    if not scan.exhausted:
        reading.update(
            detail=f"undecided: the recipient does not list this turn, and {scan.scanned} items "
                   f"did not cover the {len(presence.seen)} turns begun since the send, so the "
                   f"token's absence is not shown",
            undecided=TOKEN_SCAN_BOUNDED,
        )
        return reading
    reading.update(
        finding=HOST_LOST_TURN,
        detail=f"the recipient's turn list has no such turn ({presence.stop} after "
               f"{presence.scanned} turns) and this attempt's token is not among the "
               f"{scan.scanned} items of the {len(presence.seen)} turns begun since the send",
    )
    return reading


def record_undecided(store, request_id, reading) -> int:
    """Name an undecided reading on its attempt, or clear the name once a reading decides.

    Written only while the attempt is still a settled dispatch, and only when the recorded value
    changes, so a tick that learns nothing new writes nothing. Returns the rows changed.
    """
    if reading["finding"] == PRESENT:
        wanted = None
    elif reading["finding"] == UNKNOWN and reading.get("undecided"):
        wanted = UNDECIDED_MARK + reading["undecided"]
    else:
        return 0
    with store.transaction() as db:
        row = db.execute(
            "SELECT internal_state, state, recipient_scan FROM attempts WHERE request_id = ?",
            (request_id,),
        ).fetchone()
        if row is None or row["internal_state"] != "settled" or row["state"] != DISPATCHED:
            return 0
        current = row["recipient_scan"]
        if wanted is None and not (current or "").startswith(UNDECIDED_MARK):
            return 0
        if wanted is not None and current == wanted:
            return 0
        return db.execute(
            "UPDATE attempts SET recipient_scan = ? WHERE request_id = ?", (wanted, request_id)
        ).rowcount


def settle(store, clock, request_id, reading, *, observation=None) -> dict:
    """Record the loss and put the obligation back, in one guarded transaction.

    The delivery moves only while it is still a completion dispatched on THIS attempt, unheld and
    unacknowledged; the attempt only while it is still settled as dispatched. Anything else means
    somebody moved first - a recorded acknowledgement, a later attempt - and nothing is written.
    An earlier attempt of the same event already lost to the host turns the requeue into a hold:
    one redelivery per obligation, never a third send.
    """
    now_iso = clock.iso()
    try:
        with store.transaction() as db:
            attempt = db.execute(
                "SELECT * FROM attempts WHERE request_id = ?", (request_id,)
            ).fetchone()
            earlier = db.execute(
                "SELECT COUNT(*) AS c FROM attempts"
                " WHERE event_id = ? AND state = ? AND request_id <> ?",
                (attempt["event_id"], HOST_LOST_TURN, request_id),
            ).fetchone()["c"]
            hold = HOST_LOST_TURN if earlier else None
            moved = db.execute(
                "UPDATE deliveries SET state = ?, hold_reason = ?, next_eligible_at = NULL,"
                " dispatch_evidence = ?, lease_owner = NULL, lease_until = NULL, updated_at = ?"
                " WHERE event_id = ? AND kind = ? AND state = ? AND attempt_count = ?"
                "   AND hold_reason IS NULL"
                "   AND NOT EXISTS (SELECT 1 FROM acks k WHERE k.event_id = deliveries.event_id)",
                (QUEUED, hold, HOST_LOST_TURN, now_iso, attempt["event_id"], COMPLETION,
                 DISPATCHED, attempt["attempt_no"]),
            ).rowcount
            if moved != 1:
                return {"redelivery": NOT_MOVED,
                        "redeliveryDetail": "the delivery is no longer this attempt's unheld,"
                                            " unacknowledged dispatch"}
            record = json.loads(attempt["record"])
            record["reconciliation"] = {
                "operationReceiptChecked": observation is not None,
                "recipientTurnsChecked": True,
                # What the transport showed, unchanged: turn/start returned this turn id. The
                # loss is the attempt's state, not a rewrite of its evidence.
                "affirmativeEvidence": "receipt_turn_id",
                "checkedAt": now_iso,
            }
            assert_attempt_invariants(record)
            marked = db.execute(
                "UPDATE attempts SET state = ?, record = ?,"
                " operation_observation = COALESCE(?, operation_observation),"
                " recipient_scan = ?, affirmative_evidence = ?, reconciled_at = ?"
                " WHERE request_id = ? AND internal_state = 'settled' AND state = ?",
                (HOST_LOST_TURN, json.dumps(record), observation, reading["detail"],
                 "receipt_turn_id", now_iso, request_id, DISPATCHED),
            ).rowcount
            if marked != 1:
                raise _Raced()
            store.journal(
                HOST_LOST_TURN, attempt["event_id"],
                {"requestId": request_id, "turnId": reading["turnId"],
                 "redelivery": HELD if hold else REQUEUED, "detail": reading["detail"]},
                at=now_iso,
            )
    except _Raced:
        return {"redelivery": NOT_MOVED,
                "redeliveryDetail": "the attempt changed while the loss was being recorded"}
    return {"state": HOST_LOST_TURN, "redelivery": HELD if hold else REQUEUED,
            "holdReason": hold, "record": record}


_AWAITING_SQL = (
    "SELECT d.event_id, a.request_id FROM deliveries d"
    " JOIN attempts a ON a.event_id = d.event_id AND a.attempt_no = d.attempt_count"
    " WHERE d.kind = ? AND d.state = ? AND d.hold_reason IS NULL"
    "   AND a.internal_state = 'settled' AND a.state = ?"
    "   AND NOT EXISTS (SELECT 1 FROM acks k WHERE k.event_id = d.event_id)"
)


def awaiting_ack(store, *, after=None, limit):
    """Delivered completions still owed an acknowledgement, keyed by event, one page.

    The attempt is the delivery's CURRENT one, joined on the event as well as the number: every
    event's first attempt is number 1. A held, acknowledged or already-lost row is not a candidate.
    """
    sql = _AWAITING_SQL
    params = [COMPLETION, DISPATCHED, DISPATCHED]
    if after is not None:
        sql += " AND d.event_id > ?"
        params.append(after)
    sql += " ORDER BY d.event_id LIMIT ?"
    params.append(limit)
    return store.all(sql, tuple(params))


def still_awaiting(store, request_ids):
    """Which of these request ids are still a candidate of awaiting_ack, by the same test.

    The caller bounds the list; the request id is the attempts table's key.
    """
    request_ids = list(request_ids)
    if not request_ids:
        return set()
    marks = ", ".join("?" for _ in request_ids)
    rows = store.all(_AWAITING_SQL + f" AND a.request_id IN ({marks})",
                     (COMPLETION, DISPATCHED, DISPATCHED, *request_ids))
    return {row["request_id"] for row in rows}

File: internal/relay/delivery/hostloss.go
package delivery

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"slices"
	"strings"
	"time"

	"github.com/thisisjun786/codex-relay-workflow/internal/relay/store"
)

// Host-loss words (hostloss.py).
const (
	Present           = "present"
	Unknown           = "unknown"
	tokenScanLimit    = 200
	Requeued          = "queued"
	HeldRedelivery    = "held"
	NotMoved          = "not_moved"
	ReportOnly        = "report_only"
	ListingBoundedW   = "listing_bounded"
	ListingEmptyW     = "listing_empty"
	TokenScanBounded  = "token_scan_bounded"
	TokenWithoutTurn  = "token_without_turn"
	TokenInOtherItem  = "token_in_other_item"
	NoSendTime        = "no_send_time"
	NoTurn            = "no_turn"
	ReceiptUnsettled  = "receipt_unsettled"
	ReceiptMissing    = "receipt_missing"
	SettledReceipt    = "settled"
	UnsettledReceipt  = "unsettled"
	MissingReceipt    = "missing"
	undecidedMark     = TurnCheckUndecided + ":"
	unknownLostMark   = UnknownSendLost + ":no_trace"
	unknownUndecided  = UnknownSendUndecided + ":"
	unknownMarkPrefix = "unknown_send_"
)

var terminalTurn = []string{"completed", "interrupted", "failed"}
var deliveredAttemptStates = []string{Dispatched, HeldUncertain}

func epoch(stamp string) (float64, bool) {
	t, err := time.Parse("2006-01-02T15:04:05.999999-07:00", stamp)
	if err != nil {
		return 0, false
	}
	return float64(t.UnixMicro()) / 1e6, true
}

// Reading is what a recipient's own turns and items say about an attempt.
type Reading = Obj

func newReading(turnID any) Reading {
	return Obj{{Key: "turnId", Value: turnID}, {Key: "finding", Value: Unknown}, {Key: "status", Value: nil}, {Key: "detail", Value: nil}, {Key: "undecided", Value: nil}}
}

func allowance() float64 { return TurnStartPrecisionSeconds + DispatchTurnSkewSeconds }

// ReadRecipientTurn is hostloss.read_recipient_turn. Reads only.
func ReadRecipientTurn(adapter Adapter, clock Clock, attempt, delivery Row, turnID any) Reading {
	r := newReading(turnID)
	thread := delivery.S("recipient_thread_id")
	turn, _ := turnID.(string)
	if turn == "" {
		return set(set(r, "detail", "the attempt names no turn"), "undecided", NoTurn)
	}
	sentAt, ok := epoch(attempt.S("sent_at"))
	if !ok {
		return set(set(r, "detail", "the attempt has no send time, so absence cannot be bounded"), "undecided", NoSendTime)
	}
	allow := allowance()
	presence, err := adapter.FindDispatchedTurn(thread, turn, sentAt)
	var bounded *ListingBounded
	var empty *ListingEmpty
	switch {
	case errors.As(err, &bounded):
		return set(set(r, "detail", "undecided: "+bounded.Message), "undecided", ListingBoundedW)
	case errors.As(err, &empty):
		if clock.Now() < sentAt+allow {
			return set(r, "detail", fmt.Sprintf("the recipient lists no turns yet, and the send is less than %.0f s old: %s", allow, empty.Message))
		}
		return set(set(r, "detail", "undecided: "+empty.Message), "undecided", ListingEmptyW)
	case err != nil:
		return set(r, "detail", "unreadable: "+errorLabel(err))
	}
	listed := presence.Finding == TurnPresent
	where := "does not list this turn"
	if listed {
		status := presence.Turn.Status
		if !slices.Contains(terminalTurn, status) {
			return set(set(set(r, "finding", Present), "status", status), "detail", fmt.Sprintf("the recipient lists this turn (%s)", status))
		}
		own, err := adapter.FindTokenInTurn(thread, attempt.S("request_id"), turn, InTurnItemsMax)
		if err != nil {
			return set(r, "detail", fmt.Sprintf("unreadable: the recipient lists this turn (%s), but its items could not be read for this attempt's message: %s", status, errorLabel(err)))
		}
		if own.Found {
			return set(set(set(r, "finding", Present), "status", status), "detail", fmt.Sprintf("the recipient lists this turn (%s) with this attempt's message in it", status))
		}
		where = fmt.Sprintf("lists this turn %s without this attempt's message among its first %d items", status, own.Scanned)
	}
	if clock.Now() < sentAt+allow {
		if listed {
			return set(r, "detail", fmt.Sprintf("the recipient %s, but the send is less than %.0f s old, too recent to call the turn lost", where, allow))
		}
		return set(r, "detail", fmt.Sprintf("the recipient does not list this turn yet (%s), but the send is less than %.0f s old, too recent to call the turn lost", presence.Stop, allow))
	}
	scan, err := adapter.FindTokenSince(thread, attempt.S("request_id"), presence.Older, tokenScanLimit)
	if err != nil {
		return set(r, "detail", fmt.Sprintf("the recipient %s, and its items could not be read for this attempt's token: %s", where, errorLabel(err)))
	}
	if scan.Found {
		if listed {
			return set(set(r, "finding", Present), "detail", fmt.Sprintf("the recipient %s, but this attempt's message is in its items (turn %s)", where, pyStr(scan.TurnID)))
		}
		return set(set(set(r, "finding", Present), "detail", fmt.Sprintf("the recipient does not list this turn, but this attempt's token is in its items (turn %s)", pyStr(scan.TurnID))), "undecided", TokenWithoutTurn)
	}
	if scan.OtherKind != nil {
		return set(set(r, "detail", fmt.Sprintf("undecided: the recipient %s, and this attempt's token is in its items only in an item of type %s (turn %s), which is neither the delivered message nor agent output; not sent again", where, pyStr(scan.OtherKind), pyStr(scan.OtherTurn))), "undecided", TokenInOtherItem)
	}
	if !scan.Exhausted {
		return set(set(r, "detail", fmt.Sprintf("undecided: the recipient %s, and %d items did not reach history older than the send, so the token's absence is not shown", where, scan.Scanned)), "undecided", TokenScanBounded)
	}
	if listed {
		return set(set(r, "finding", HostLostTurn), "detail", fmt.Sprintf("the recipient %s, and this attempt's token is not among the %d items since the send (%d listed turns begun since it): the host lost the turn's content", where, scan.Scanned, len(presence.Seen)))
	}
	return set(set(r, "finding", HostLostTurn), "detail", fmt.Sprintf("the recipient's turn list has no such turn (%s after %d turns) and this attempt's token is not among the %d items since the send (%d listed turns begun since it)", presence.Stop, presence.Scanned, scan.Scanned, len(presence.Seen)))
}

// RecordUndecided is hostloss.record_undecided; returns the rows changed.
func RecordUndecided(ctx context.Context, s *store.Store, requestID string, reading Reading) (int64, error) {
	var wanted any
	if u, _ := get(reading, "undecided"); truthy(u) {
		wanted = undecidedMark + pyStr(u)
	} else if str(reading, "finding") != Present {
		return 0, nil
	}
	var changed int64
	err := s.Transaction(ctx, func(ctx context.Context, _ *sql.Conn) error {
		row, err := one(ctx, s, "SELECT internal_state, state, recipient_scan FROM attempts WHERE request_id = ?", requestID)
		if err != nil || row == nil || row.S("internal_state") != "settled" || !slices.Contains(deliveredAttemptStates, row.S("state")) {
			return err
		}
		current := row.S("recipient_scan")
		if wanted == nil && !strings.HasPrefix(current, undecidedMark) {
			return nil
		}
		if wanted != nil && !row.N("recipient_scan") && current == wanted {
			return nil
		}
		changed, err = execSQL(ctx, s, "UPDATE attempts SET recipient_scan = ? WHERE request_id = ?", wanted, requestID)
		return err
	})
	return changed, err
}

var errRaced = errors.New("raced")

// SettleLoss is hostloss.settle: record the loss and put the obligation back, once.
func SettleLoss(ctx context.Context, s *store.Store, clock Clock, requestID string, reading Reading, observation any) (Obj, error) {
	now := clock.ISO()
	var record Obj
	var hold any
	notMoved := false
	err := s.Transaction(ctx, func(ctx context.Context, _ *sql.Conn) error {
		attempt, err := one(ctx, s, "SELECT * FROM attempts WHERE request_id = ?", requestID)
		if err != nil {
			return err
		}
		earlier, err := one(ctx, s, "SELECT COUNT(*) AS c FROM attempts WHERE event_id = ? AND state = ? AND request_id <> ?", attempt.S("event_id"), HostLostTurn, requestID)
		if err != nil {
			return err
		}
		if earlier.I("c") > 0 {
			hold = HostLostTurn
		}
		moved, err := execSQL(ctx, s, "UPDATE deliveries SET state = ?, hold_reason = ?, next_eligible_at = NULL, dispatch_evidence = ?, lease_owner = NULL, lease_until = NULL, updated_at = ? WHERE event_id = ? AND kind = ? AND state = ? AND attempt_count = ? AND hold_reason IS NULL AND NOT EXISTS (SELECT 1 FROM acks k WHERE k.event_id = deliveries.event_id AND (k.verified = 'verified' OR ? IS NULL OR k.ack_at >= ?))",
			Queued, hold, HostLostTurn, now, attempt.S("event_id"), Completion, Dispatched, attempt.I("attempt_no"), attempt.Opt("sent_at"), attempt.Opt("sent_at"))
		if err != nil {
			return err
		}
		if moved != 1 {
			notMoved = true
			return nil
		}
		record = loadsObj(attempt.S("record"))
		evidence := attempt.S("affirmative_evidence")
		if evidence == "" {
			evidence = "receipt_turn_id"
		}
		record = set(record, "reconciliation", Obj{{Key: "operationReceiptChecked", Value: observation != nil}, {Key: "recipientTurnsChecked", Value: true}, {Key: "affirmativeEvidence", Value: evidence}, {Key: "checkedAt", Value: now}})
		if err := AssertAttemptInvariants(record); err != nil {
			return err
		}
		marked, err := execSQL(ctx, s, "UPDATE attempts SET state = ?, record = ?, operation_observation = COALESCE(?, operation_observation), recipient_scan = ?, affirmative_evidence = ?, reconciled_at = ? WHERE request_id = ? AND internal_state = 'settled' AND state IN (?, ?)",
			HostLostTurn, dumps(record), observation, str(reading, "detail"), evidence, now, requestID, Dispatched, HeldUncertain)
		if err != nil {
			return err
		}
		if marked != 1 {
			return errRaced
		}
		redelivery := Requeued
		if hold != nil {
			redelivery = HeldRedelivery
		}
		turnID, _ := get(reading, "turnId")
		return journal(ctx, s, HostLostTurn, attempt.S("event_id"), Obj{{Key: "requestId", Value: requestID}, {Key: "turnId", Value: turnID}, {Key: "redelivery", Value: redelivery}, {Key: "detail", Value: str(reading, "detail")}}, now)
	})
	if errors.Is(err, errRaced) {
		return Obj{{Key: "redelivery", Value: NotMoved}, {Key: "redeliveryDetail", Value: "the attempt changed while the loss was being recorded"}}, nil
	}
	if err != nil {
		return nil, err
	}
	if notMoved {
		return Obj{{Key: "redelivery", Value: NotMoved}, {Key: "redeliveryDetail", Value: "the delivery is no longer this attempt's unheld, unacknowledged dispatch"}}, nil
	}
	redelivery := Requeued
	if hold != nil {
		redelivery = HeldRedelivery
	}
	return Obj{{Key: "state", Value: HostLostTurn}, {Key: "redelivery", Value: redelivery}, {Key: "holdReason", Value: hold}, {Key: "record", Value: record}}, nil
}

func foldCandidates(p TurnPresence, sentAt float64) []TurnInfo {
	var out []TurnInfo
	for _, t := range p.SeenTurns {
		if t.StartedAt != nil && *t.StartedAt <= sentAt+TurnStartPrecisionSeconds {
			out = append(out, t)
		}
	}
	if p.StopTurn != nil {
		out = append(out, *p.StopTurn)
	}
	return out
}

// ReadUnknownSend is hostloss.read_unknown_send. Reads only; never allows a second send.
func ReadUnknownSend(adapter Adapter, clock Clock, attempt, delivery Row, receipt string) Reading {
	r := Obj{{Key: "turnId", Value: nil}, {Key: "finding", Value: Unknown}, {Key: "detail", Value: nil}, {Key: "undecided", Value: nil}, {Key: "pending", Value: false}}
	thread := delivery.S("recipient_thread_id")
	request := attempt.S("request_id")
	sentAt, ok := epoch(attempt.S("sent_at"))
	if !ok {
		return set(set(r, "detail", "the attempt has no send time, so absence cannot be bounded"), "undecided", NoSendTime)
	}
	allow := allowance()
	pending := func(detail string) Reading { return set(set(r, "detail", detail), "pending", true) }
	undecided := func(detail, why string) Reading { return set(set(r, "detail", detail), "undecided", why) }
	if clock.Now() < sentAt+allow {
		return pending(fmt.Sprintf("the send is less than %.0f s old, too recent to call it lost; read again", allow))
	}
	presence, err := adapter.FindDispatchedTurn(thread, "", sentAt)
	var bounded *ListingBounded
	var empty *ListingEmpty
	switch {
	case errors.As(err, &bounded):
		return undecided("undecided: "+bounded.Message, ListingBoundedW)
	case errors.As(err, &empty):
		return undecided("undecided: "+empty.Message, ListingEmptyW)
	case err != nil:
		return pending("unreadable: " + errorLabel(err))
	}
	folded := foldCandidates(presence, sentAt)
	var running []string
	for _, t := range presence.SeenTurns {
		if !slices.Contains(terminalTurn, t.Status) {
			running = append(running, t.TurnID)
		}
	}
	if presence.StopTurn != nil && !slices.Contains(terminalTurn, presence.StopTurn.Status) {
		running = append(running, presence.StopTurn.TurnID)
	}
	if len(running) > 0 {
		return pending(fmt.Sprintf("a turn begun since the send is still running (%s); read again once it ends", running[0]))
	}
	scan, err := adapter.FindTokenSince(thread, request, presence.Older, tokenScanLimit)
	if err != nil {
		return pending("unreadable: the recipient's items could not be read for this attempt's token: " + errorLabel(err))
	}
	if scan.Found {
		return set(set(set(r, "finding", Present), "turnId", scan.TurnID), "detail", fmt.Sprintf("this attempt's message is in the recipient's items since the send (turn %s, %d items read)", pyStr(scan.TurnID), scan.Scanned))
	}
	type owned struct {
		turn TurnInfo
		scan TokenScan
	}
	var owns []owned
	for _, turn := range folded {
		own, err := adapter.FindTokenInTurn(thread, request, turn.TurnID, InTurnItemsMax)
		if err != nil {
			return pending(fmt.Sprintf("unreadable: the items of turn %s, one the send could have been folded into, could not be read: %s", turn.TurnID, errorLabel(err)))
		}
		if own.Found {
			return set(set(set(r, "finding", Present), "turnId", turn.TurnID), "detail", fmt.Sprintf("this attempt's message is in turn %s, begun by the send, which the send was folded into", turn.TurnID))
		}
		owns = append(owns, owned{turn, own})
	}
	for _, o := range owns {
		if o.scan.OtherKind != nil {
			return undecided(fmt.Sprintf("undecided: this attempt's token is in turn %s, which the send could have been folded into, only in an item of type %s, which is neither the delivered message nor agent output; not sent again", o.turn.TurnID, pyStr(o.scan.OtherKind)), TokenInOtherItem)
		}
	}
	if scan.OtherKind != nil {
		return undecided(fmt.Sprintf("undecided: this attempt's token is in the recipient's items only in an item of type %s (turn %s), which is neither the delivered message nor agent output; not sent again", pyStr(scan.OtherKind), pyStr(scan.OtherTurn)), TokenInOtherItem)
	}
	if !scan.Exhausted {
		return undecided(fmt.Sprintf("undecided: %d items did not reach history older than the send, so the token's absence is not shown", scan.Scanned), TokenScanBounded)
	}
	for _, o := range owns {
		if !o.scan.Exhausted {
			return undecided(fmt.Sprintf("undecided: turn %s, begun by the send, could have taken it, and its first %d items did not reach its end", o.turn.TurnID, o.scan.Scanned), TokenScanBounded)
		}
	}
	if receipt != SettledReceipt {
		said := "the transport has not settled this request's receipt (in_progress_or_unknown)"
		why := ReceiptUnsettled
		if receipt == MissingReceipt {
			said, why = "the transport holds no receipt for this request id", ReceiptMissing
		}
		return undecided("undecided: the recipient keeps no trace of this send, but "+said, why)
	}
	foldedText := ""
	if len(folded) > 0 {
		ids := make([]string, len(folded))
		for i, t := range folded {
			ids[i] = t.TurnID
		}
		foldedText = " nor in " + strings.Join(ids, ", ") + ", the turns begun by it"
	}
	return set(set(r, "finding", UnknownSendLost), "detail", fmt.Sprintf("the recipient keeps no trace of this send: its turn list (%s after %d turns) shows %d turns begun since it, none running, and this attempt's token is not among the %d items since it%s; held for the parent, never sent again", presence.Stop, presence.Scanned, len(presence.Seen), scan.Scanned, foldedText))
}
